core/src/main/java/org/jclouds/rest/internal/RestAnnotationProcessor.java

```diff
@@ -357,6 +357,22 @@ public GeneratedHttpRequest apply(Invocation invocation) {
       if (request.getPayload() != null) {
          contentMetadataCodec.fromHeaders(request.getPayload().getContentMetadata(), headers);
       }
+
+      boolean isEmptyBody = false;
+      if ("POST".equals(requestMethod) || "PUT".equals(requestMethod)) {
+         if (request.getPayload() != null) {
+            Long length = request.getPayload().getContentMetadata().getContentLength();
+            if (length != null && length == 0) {
+               isEmptyBody = true;
+            }
+         } else {
+            isEmptyBody = true;
+         }
+      }
+      if (isEmptyBody) {
+         request = request.toBuilder().removeHeader("Expect").build();
+      }
```

Comment:
@nacx Refactored the code to a separate method and added a unit test case to check the functionality.